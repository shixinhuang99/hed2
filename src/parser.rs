use std::{collections::HashSet, net::IpAddr};

use indexmap::IndexMap;
use serde::{Deserialize, Serialize};

const SYSTEM_GROUP_NAME: &str = "System";

#[derive(Debug, Clone)]
pub enum Line {
	Valid {
		ip: String,
		hosts: Vec<String>,
		enabled: bool,
	},
	Empty,
	Other(String),
	Group(String),
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Item {
	ip: String,
	hosts: Vec<Host>,
	group: String,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Host {
	content: String,
	enabled: bool,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Group {
	name: String,
	text: String,
	list: Vec<Item>,
}

pub fn is_ip(s: &str) -> bool {
	s.parse::<IpAddr>().is_ok()
}

fn key(s1: &str, s2: &str) -> String {
	format!("{}_{}", s1, s2)
}

fn text_to_lines(text: &str) -> Vec<Line> {
	let mut lines = vec![];

	for line in text.lines() {
		let tokens = line.split_whitespace().collect::<Vec<&str>>();

		if tokens.is_empty() {
			lines.push(Line::Empty);
			continue;
		}

		if tokens.len() == 1 {
			if let Some(stripped) = tokens[0].strip_prefix("#[") {
				if let Some(group) = stripped.strip_suffix("]") {
					lines.push(Line::Group(group.trim().to_string()));
					continue;
				}
			}
		}

		let enabled = tokens[0] != "#";
		let may_ip = if enabled {
			tokens[0]
		} else {
			tokens[1]
		};

		if is_ip(may_ip) {
			let mut hosts = vec![];
			let skip = if enabled {
				1
			} else {
				2
			};

			for host in tokens.iter().skip(skip) {
				if let Some(idx) = host.find('#') {
					let slice = &host[0..idx];
					if !slice.is_empty() {
						hosts.push(slice.to_string());
					}
					break;
				}
				hosts.push(host.to_string());
			}

			if !hosts.is_empty() {
				lines.push(Line::Valid {
					ip: may_ip.to_string(),
					hosts,
					enabled,
				});
				continue;
			}
		}

		lines.push(Line::Other(line.trim().to_string()));
	}

	lines
}

fn lines_to_list(lines: &[Line], group: Option<&str>) -> Vec<Item> {
	let mut item_map: IndexMap<String, Item> = IndexMap::new();
	let mut current_group = group.unwrap_or(SYSTEM_GROUP_NAME);
	let mut tmp_map: IndexMap<String, Item> = IndexMap::new();

	for line in lines {
		match line {
			Line::Valid { ip, hosts, enabled } => {
				if hosts.is_empty() {
					continue;
				}

				let mut hosts = hosts
					.iter()
					.map(|content| Host {
						content: content.clone(),
						enabled: *enabled,
					})
					.collect::<Vec<Host>>();

				let key = key(current_group, ip);

				tmp_map
					.entry(key)
					.and_modify(|item| item.hosts.append(&mut hosts))
					.or_insert(Item {
						ip: ip.clone(),
						hosts,
						group: current_group.to_string(),
					});
			}
			Line::Group(name) => {
				if group.is_some()
					|| name == SYSTEM_GROUP_NAME
					|| current_group != SYSTEM_GROUP_NAME
						&& current_group != name
				{
					continue;
				}
				let t = tmp_map.clone();
				tmp_map.clear();
				for (k, mut v) in t {
					item_map
						.entry(k)
						.and_modify(|item| item.hosts.append(&mut v.hosts))
						.or_insert(v);
				}
				if current_group == name {
					current_group = SYSTEM_GROUP_NAME;
				} else {
					current_group = name;
				}
			}
			_ => (),
		}
	}

	if !tmp_map.is_empty() {
		for (k, mut v) in tmp_map {
			if group.is_none() {
				v.group = SYSTEM_GROUP_NAME.to_string();
			}
			item_map
				.entry(k)
				.and_modify(|item| item.hosts.append(&mut v.hosts))
				.or_insert(v);
		}
	}

	for item in item_map.values_mut() {
		let mut set: HashSet<String> = HashSet::new();

		item.hosts.retain(|host| {
			if set.contains(&host.content) {
				return false;
			}
			set.insert(host.content.clone());
			true
		});
	}

	item_map.into_values().collect()
}

pub fn text_to_groups(text: String) -> Vec<Group> {
	let lines = text_to_lines(&text);
	let mut group_map: IndexMap<String, Group> = IndexMap::new();
	let mut current_group: Option<&String> = None;
	let mut group_lines: Vec<&Line> = vec![];

	for line in &lines {
		if let Some(cur_group_name) = current_group {
			if let Line::Group(new_group_name) = line {
				if new_group_name == cur_group_name {
					let owned_lines = group_lines
						.clone()
						.into_iter()
						.cloned()
						.collect::<Vec<Line>>();
					let text = lines_to_text(&owned_lines);
					group_map
						.entry(new_group_name.clone())
						.and_modify(|group| {
							group.text = format!("{}\n{}", group.text, text);
						})
						.or_insert(Group {
							name: new_group_name.clone(),
							text,
							list: vec![],
						});
					group_lines.clear();
					current_group = None;
					continue;
				}
			}
			group_lines.push(line);
		} else if let Line::Group(name) = line {
			if name != SYSTEM_GROUP_NAME {
				current_group = Some(name);
			}
		}
	}

	let system_group = Group {
		name: SYSTEM_GROUP_NAME.to_string(),
		text,
		list: lines_to_list(&lines, None),
	};
	for item in &system_group.list {
		if let Some(group) = group_map.get_mut(&item.group) {
			group.list.push(item.clone());
		}
	}
	let mut groups = vec![system_group];
	groups.append(&mut group_map.into_values().collect::<Vec<Group>>());
	groups
}

fn hosts_partition(hosts: Vec<Host>) -> (Vec<Host>, Vec<Host>) {
	hosts.into_iter().partition(|host| host.enabled)
}

fn hosts_to_lines(hosts: Vec<Host>, ip: &str, enabled: bool) -> Vec<Line> {
	hosts
		.chunks(10)
		.map(|chunk| Line::Valid {
			ip: ip.to_string(),
			hosts: chunk.iter().cloned().map(|host| host.content).collect(),
			enabled,
		})
		.collect()
}

fn list_to_lines(list: Vec<Item>, old_lines: Vec<Line>) -> Vec<Line> {
	let mut lines = vec![];
	let mut hosts_map = list
		.into_iter()
		.map(|item| (item.ip, item.hosts))
		.collect::<IndexMap<String, Vec<Host>>>();

	for line in old_lines {
		if let Line::Valid { ip, .. } = line {
			if let Some(hosts) = hosts_map.get_mut(&ip) {
				if hosts.is_empty() {
					lines.push(Line::Empty);
					continue;
				}
				let mut new_hosts = vec![];
				new_hosts.append(hosts);
				let (enabled, disabled) = hosts_partition(new_hosts);
				lines.append(&mut hosts_to_lines(enabled, &ip, true));
				lines.append(&mut hosts_to_lines(disabled, &ip, false));
			} else {
				lines.push(Line::Empty);
			}
		} else {
			lines.push(line);
		}
	}

	for (ip, hosts) in hosts_map {
		if hosts.is_empty() {
			continue;
		}
		lines.push(Line::Empty);
		let (enabled, disabled) = hosts_partition(hosts);
		lines.append(&mut hosts_to_lines(enabled, &ip, true));
		lines.append(&mut hosts_to_lines(disabled, &ip, false));
	}

	lines
}

fn lines_to_text(lines: &[Line]) -> String {
	let mut text_lines = vec![];

	let mut is_pre_empty = lines
		.first()
		.is_some_and(|line| matches!(line, Line::Empty));

	for line in lines {
		match line {
			Line::Valid { ip, hosts, enabled } => {
				let mut vs = vec![];
				if !enabled {
					vs.push("#".to_string());
				}
				vs.push(ip.clone());
				vs.append(&mut hosts.clone());
				text_lines.push(vs.join(" "));
			}
			Line::Other(s) => {
				text_lines.push(s.clone());
			}
			Line::Empty => {
				if !is_pre_empty {
					text_lines.push(String::new());
				}
			}
			Line::Group(name) => {
				text_lines.push(format!("#[{}]", name));
			}
		};
		is_pre_empty = matches!(line, Line::Empty);
	}

	if !is_pre_empty {
		text_lines.push(String::new());
	}

	text_lines.join("\n")
}

pub fn list_to_text(list: Vec<Item>, old_text: String) -> String {
	let lines = text_to_lines(&old_text);
	let new_lines = list_to_lines(list, lines);
	lines_to_text(&new_lines)
}

#[cfg(test)]
mod tests {
	use insta::{assert_debug_snapshot, assert_snapshot};

	use super::{
		lines_to_list, lines_to_text, list_to_lines, text_to_groups,
		text_to_lines, Host, Item,
	};

	#[test]
	fn test_parser() {
		static MOCK_HOSTS: &str = include_str!("../fixture/hosts");

		let lines = text_to_lines(MOCK_HOSTS);
		assert_debug_snapshot!("text_to_lines", lines);

		let mut list = lines_to_list(&lines, None);
		assert_debug_snapshot!("lines_to_list", list);

		for item in &mut list {
			if item.ip == "1.1.1.1" {
				item.hosts.pop();
				item.hosts.push(Host {
					content: "o.com".to_string(),
					enabled: true,
				});
				item.hosts.push(Host {
					content: "13.com".to_string(),
					enabled: false,
				});
			} else if item.ip == "3.3.3.3" && item.group == "foo" {
				item.hosts.push(Host {
					content: "d.com".to_string(),
					enabled: true,
				})
			}
		}

		list.push(Item {
			ip: "2.2.2.2".to_string(),
			hosts: vec![
				Host {
					content: "foo.com".to_string(),
					enabled: true,
				},
				Host {
					content: "bar.com".to_string(),
					enabled: false,
				},
			],
			group: "Another Group".to_string(),
		});

		let new_lines = list_to_lines(list.clone(), lines.clone());
		assert_debug_snapshot!("list_to_lines", new_lines);

		let text = lines_to_text(&new_lines);
		assert_snapshot!("lines_to_text", text);

		let groups = text_to_groups(MOCK_HOSTS.to_string());
		assert_debug_snapshot!("text_to_groups", groups);
	}
}
