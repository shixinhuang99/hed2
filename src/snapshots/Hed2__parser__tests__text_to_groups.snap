---
source: src/parser.rs
expression: groups
snapshot_kind: text
---
[
    Group {
        name: "System",
        text: "\n\n\n\n   #    comment  \n               \n#    1.1.1.1 1.com 2.com 3.com 4.com 5.com 6.com 7.com 8.com 9.com 10.com\n\n1.1.1.1 a.com    c.com     b.com f.com   #     ##     comment \n1.1.1.1   e.com h.com  g.com i.com  k.com j.com l.com m.com   n.com#     ##     comment \n\n# comment\n# comment\n# comment\n# comment\n# comment\n\n# 1.1.1.1 11.com 12.com\n\n\n           2402:1200:4f00:1234:0000:5678:9abc:def0   b.com  \n\n\n4.4.4.4    \n# 4.4.4.4 \n\n#[foo]    \n\n#[bar]\n  3.3.3.3 a.com\n\n  #   3.3.3.3 b.com\n#[bar]\n\n#[foo]      \n\n#abcdefg\n\n#[foo]\n3.3.3.3 e.com\n5.5.5.5 a.com\n#[foo]\n\n#[System]\n3.3.3.3 c.com\n#[System]\n\n#[baz]\n5.5.5.5 a.com\n\n#[foo]\n5.5.5.5 c.com\n#[foo]\n",
        list: [
            Item {
                ip: "1.1.1.1",
                hosts: [
                    Host {
                        content: "1.com",
                        enabled: false,
                    },
                    Host {
                        content: "2.com",
                        enabled: false,
                    },
                    Host {
                        content: "3.com",
                        enabled: false,
                    },
                    Host {
                        content: "4.com",
                        enabled: false,
                    },
                    Host {
                        content: "5.com",
                        enabled: false,
                    },
                    Host {
                        content: "6.com",
                        enabled: false,
                    },
                    Host {
                        content: "7.com",
                        enabled: false,
                    },
                    Host {
                        content: "8.com",
                        enabled: false,
                    },
                    Host {
                        content: "9.com",
                        enabled: false,
                    },
                    Host {
                        content: "10.com",
                        enabled: false,
                    },
                    Host {
                        content: "a.com",
                        enabled: true,
                    },
                    Host {
                        content: "c.com",
                        enabled: true,
                    },
                    Host {
                        content: "b.com",
                        enabled: true,
                    },
                    Host {
                        content: "f.com",
                        enabled: true,
                    },
                    Host {
                        content: "e.com",
                        enabled: true,
                    },
                    Host {
                        content: "h.com",
                        enabled: true,
                    },
                    Host {
                        content: "g.com",
                        enabled: true,
                    },
                    Host {
                        content: "i.com",
                        enabled: true,
                    },
                    Host {
                        content: "k.com",
                        enabled: true,
                    },
                    Host {
                        content: "j.com",
                        enabled: true,
                    },
                    Host {
                        content: "l.com",
                        enabled: true,
                    },
                    Host {
                        content: "m.com",
                        enabled: true,
                    },
                    Host {
                        content: "n.com",
                        enabled: true,
                    },
                    Host {
                        content: "11.com",
                        enabled: false,
                    },
                    Host {
                        content: "12.com",
                        enabled: false,
                    },
                ],
                group: "System",
            },
            Item {
                ip: "2402:1200:4f00:1234:0000:5678:9abc:def0",
                hosts: [
                    Host {
                        content: "b.com",
                        enabled: true,
                    },
                ],
                group: "System",
            },
            Item {
                ip: "3.3.3.3",
                hosts: [
                    Host {
                        content: "a.com",
                        enabled: true,
                    },
                    Host {
                        content: "b.com",
                        enabled: false,
                    },
                    Host {
                        content: "e.com",
                        enabled: true,
                    },
                ],
                group: "foo",
            },
            Item {
                ip: "5.5.5.5",
                hosts: [
                    Host {
                        content: "a.com",
                        enabled: true,
                    },
                ],
                group: "foo",
            },
            Item {
                ip: "3.3.3.3",
                hosts: [
                    Host {
                        content: "c.com",
                        enabled: true,
                    },
                ],
                group: "System",
            },
            Item {
                ip: "5.5.5.5",
                hosts: [
                    Host {
                        content: "a.com",
                        enabled: true,
                    },
                    Host {
                        content: "c.com",
                        enabled: true,
                    },
                ],
                group: "System",
            },
        ],
    },
    Group {
        name: "foo",
        text: "#[bar]\n3.3.3.3 a.com\n\n# 3.3.3.3 b.com\n#[bar]\n\n3.3.3.3 e.com\n5.5.5.5 a.com\n",
        list: [
            Item {
                ip: "3.3.3.3",
                hosts: [
                    Host {
                        content: "a.com",
                        enabled: true,
                    },
                    Host {
                        content: "b.com",
                        enabled: false,
                    },
                    Host {
                        content: "e.com",
                        enabled: true,
                    },
                ],
                group: "foo",
            },
            Item {
                ip: "5.5.5.5",
                hosts: [
                    Host {
                        content: "a.com",
                        enabled: true,
                    },
                ],
                group: "foo",
            },
        ],
    },
]
